js/core/src/registry.ts

```diff
@@ -133,15 +139,38 @@ export class Registry {
     this.actionsById[key] = action;
   }
 
+  /**
+   * Registers an action promise in the registry.
+   */
+  registerActionAsync<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
+    type: ActionType,
+    name: string,
+    action: Promise<Action<I, O>>
+  ) {
+    const key = `/${type}/${name}`;
+    logger.debug(`registering ${key}`);
+    if (this.actionsById.hasOwnProperty(key)) {
+      // TODO: Make this an error!
+      logger.warn(
+        `WARNING: ${key} already has an entry in the registry. Overwriting.`
+      );
+    }
+    this.actionsById[key] = action;
+  }
+
   /**
    * Returns all actions in the registry.
    * @returns All actions in the registry.
    */
   async listActions(): Promise<ActionsRecord> {
     await this.initializeAllPlugins();
+    const actions: Record<string, Action<z.ZodTypeAny, z.ZodTypeAny>> = {};
+    for (let [key, action] of Object.entries(this.actionsById)) {
+      actions[key] = await action;
+    }
```

Comment:
Should this be:

```suggestion
    await Promise.all(Object.entries(this.actionsById).map(async ([key,action]) => {
      actions[key] = await action;
    });
```

so that it doesn't wait serially on the actions? Maybe it doesn't really matter since the promises are already in flight tho.